### PR TITLE
Java: convert sinks to CSV

### PR DESCRIPTION
**Pull request by tamasvajk**
~~[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/1257/)~~
~~[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/1285/)~~
[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/1295/)

### CODE REVIEW

File: java/ql/src/Security/CWE/CWE-297/UnsafeHostnameVerification.ql [hunk 1]
```diff
@@ -49,14 +50,7 @@ class TrustAllHostnameVerifierConfiguration extends DataFlow::Configuration {
...
 
-  override predicate isSink(DataFlow::Node sink) {
-    exists(MethodAccess ma, Method m |
-      (m instanceof SetDefaultHostnameVerifierMethod or m instanceof SetHostnameVerifierMethod) and
-      ma.getMethod() = m
-    |
-      ma.getArgument(0) = sink.asExpr()
-    )
-  }
+  override predicate isSink(DataFlow::Node sink) { sinkNode(sink, "set-hostname") }
```
Comment (aschackmull) [inline java/ql/src/Security/CWE/CWE-297/UnsafeHostnameVerification.ql:53]: "set-hostname" sounds like the sink is a hostname expression that gets set. Possibly change to "set-hostname-verifier"?

File: java/ql/src/semmle/code/java/security/ResponseSplitting.qll [hunk 2]
```diff
@@ -5,41 +5,30 @@ import semmle.code.java.dataflow.DataFlow
 import semmle.code.java.dataflow.FlowSources
 import semmle.code.java.frameworks.Servlets
 import semmle.code.java.frameworks.JaxWS
+private import semmle.code.java.dataflow.ExternalFlow
 
 /** A sink that is vulnerable to an HTTP header splitting attack. */
-abstract class HeaderSplittingSink extends DataFlow::Node { }
+class HeaderSplittingSink extends DataFlow::Node {
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/ResponseSplitting.qll:11]: We can't just remove the abstract class. That's an extension point for anyone using the `Customizations.qll` extension mechanism. Instead we'll need to subclass it with a `DefaultHeaderSplittingSink extends HeaderSplittingSink`.

File: java/ql/src/semmle/code/java/security/ResponseSplitting.qll [hunk 3]
```diff
@@ -5,41 +5,30 @@ import semmle.code.java.dataflow.DataFlow
...
+  HeaderSplittingSink() { sinkNode(this, "header-splitting") }
+}
+
+private class HeaderSplittingSinkModel extends SinkModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "javax.servlet.http;HttpServletResponse;false;addCookie;;;Argument[0];header-splitting",
+        "javax.servlet.http;HttpServletResponse;false;addHeader;;;Argument;header-splitting",
+        "javax.servlet.http;HttpServletResponse;false;setHeader;;;Argument;header-splitting",
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/ResponseSplitting.qll:21]: Could we add argument indices here, please?

File: java/ql/src/semmle/code/java/frameworks/google/GoogleHttpClientApi.qll [hunk 4]
```diff
@@ -18,16 +11,17 @@ private class TypeLiteralToParseAsFlowConfiguration extends DataFlow5::Configura
 
   override predicate isSource(DataFlow::Node source) { source.asExpr() instanceof TypeLiteral }
 
-  override predicate isSink(DataFlow::Node sink) {
-    exists(MethodAccess ma |
-      ma.getAnArgument() = sink.asExpr() and
-      ma.getMethod() instanceof ParseAsMethod
-    )
-  }
+  override predicate isSink(DataFlow::Node sink) { sinkNode(sink, "google-parse-as") }
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/google/GoogleHttpClientApi.qll:14]: This isn't really a query sink - I think we'll want to revert this change. Just because `GoogleHttpClientApi.qll` happens to have an internal use of the data-flow library that doesn't mean that we want to represent its sink through a csv model.

**Review by aschackmull — CHANGES_REQUESTED** [review]
As an overall comment, we're getting a bunch of opaque sink identifiers, which really could use qldoc. The best way to add qldoc to these identifiers is probably to introduce the convention that all references to `sinkNode(n, "some-sink-id")` occur as charpreds of simple wrapper classes. E.g.:
```
class XssSink extends DataFlow::Node {
  XssSink() { sinkNode(this, "xss") }
}
```
This allows us a place to introduce a nice QL class name and suitable qldoc that explains what the sink is.

File: java/ql/src/Security/CWE/CWE-319/HttpsUrls.ql [hunk 5]
```diff
@@ -30,26 +31,12 @@ class HTTPString extends StringLiteral {
...
   HTTPStringToURLOpenMethodFlowConfig() { this = "HttpsUrls::HTTPStringToURLOpenMethodFlowConfig" }
 
   override predicate isSource(DataFlow::Node src) { src.asExpr() instanceof HTTPString }
 
-  override predicate isSink(DataFlow::Node sink) {
-    exists(MethodAccess m |
-      sink.asExpr() = m.getQualifier() and m.getMethod() instanceof URLOpenMethod
-    )
-  }
+  override predicate isSink(DataFlow::Node sink) { sinkNode(sink, "open-url") }
```
Comment (aschackmull) [inline java/ql/src/Security/CWE/CWE-319/HttpsUrls.ql:39]: Could we extract this `sinkNode(..)` reference to a class as well such that it can get qldoc?

File: java/ql/src/Security/CWE/CWE-319/HttpsUrls.ql [hunk 6]
```diff
@@ -63,6 +50,13 @@ class HTTPStringToURLOpenMethodFlowConfig extends TaintTracking::Configuration {
   }
 }
 
+/**
+ * A sink that represents a URL opening method call, such as a call to `java.net.URL.openConnection()`.
+ */
+private class URLOpenSink extends DataFlow::Node {
```
Comment (aschackmull) [inline java/ql/src/Security/CWE/CWE-319/HttpsUrls.ql:56]:
```suggestion
private class UrlOpenSink extends DataFlow::Node {
```

Comment (tamasvajk) [reply java/ql/src/Security/CWE/CWE-319/HttpsUrls.ql:56]: I don't mind changing this to `UrlOpenSink`, but a couple of lines above we already have `HTTPStringToURLOpenMethodFlowConfig`, so it would look a bit strange, wouldn't it?

Comment (aschackmull) [reply java/ql/src/Security/CWE/CWE-319/HttpsUrls.ql:56]: Right, that ought to be renamed to `HttpStringToUrlOpenMethodFlowConfig`.

**Review by aschackmull — APPROVED** [review]
